uogetopt.3
.TH uogetopt 3
.SH NAME
uogetopt \- GNU getopt clone with extented functionality
.SH SYNTAX
.B #include "uogetopt.h"
.RS
.nf
typedef struct {
    char shortname;
    const char *longname;
    int argtype; /* 0 or 1 no argument, set (*arg) from val. */
    void *var; /* mandatory */
    int val; /* val for *var in case of type 0 or 1 */
    const char *help;
    const char *paraname;
} uogetopt_t;
.fi
.RE
.sp
.BI "void uogetopt(const char *prog, const char *package, const char *version, "
.br
.BI "     int *argc, char **argv,"
.br
.BI "     void (*out)(int iserr,const char *),"
.br
.BI "     const char *head,uogetopt_t *,const char *tail);"
.br 
.br 
.BI "void uogetopt_out (int iserr,const char *s);"

.SH DESCRIPTION
The 
.B uogetopt
function handles the options included in the 
argument/option vector described by the
.I argc
and 
.I argv
arguments to 
.BR uogetopt .
It provides support for the
.IR "--help" ,
.IR "--longhelp" 
and
.I "--version"
options described below, see "IMPLICIT OPTIONS".

Options are described by the type
.B "uogetopt_t"
which is described in the section named
.IR "STRUCT UOGETOPT" . 
.B uogetopt
manages the assignment of commandline option or flag values to program variables.
It includes direct support for the most often used option data types (flags, strings
and long numbers), and supports other data types through callback functions.

.B uogetopt 
handles 
.B all
options. If it detects an error it will exit (this is also true in case an implicit
option is used). 

If the environment variable
.I POSIXLY_CORRECT
is set 
.B uogetopt 
will stop at the first non-option argument and reorder the 
.I argv
array to begin at that argument (and update *argc). Otherwise
.B uogetopt 
will scan the whole 
.I argv 
array for options and will return only non-options in the argv
array.

.B uogetopt
will reorder the 
.I argv array 
and change 
.IR *argc ,
but will not change the strings in the array. If 
.I argv
members are dynamically allocated you should possibly make sure you have some
other points to them to be able to free them again (be sure to understand how
the 
.B UOGO_STRING
type works before you do that).

.SS "Description of the arguments for uogetopt"
.TP
.B "const char *prog"
is mandatory and contains the name of the program, to be used in version
strings (the --version and --help style options) and possibly for the
usage line of the help output (see the 
.I "head"
variable, below).
.TP
.B "const char *package"
is optional and holds the name of the package the programm belongs to. If it's
given then the version strings will look like this:
.br
.EX
  lrz (lrzsz) 0.12.20
.EE
.br
otherwise the will look like this:
.br
.EX
  lrz 0.12.20
.EE

.TP
.B "const char *version"
is mandatory and contains the version number of the program, to be used in version
strings (the --version and --help style options).

.TP
.B "int *argc"
must contain the number of arguments 
.B "argv"
contains, and will after return of the 
.B uogetopt
function contain the number of arguments left in the new array.
.TP
.B "char **argv"
is the argument array.
.TP
.B "void (*out)(int iserr,const char *)"
is the function called by 
.B uogetopt 
to print the output texts (help, version or errors). If this argument is
.I NULL
then an crude internal function be used (uogetopt_out), which simply sends
the text to the standard or standard error output.
.TP
.B "const char *head"
is optional. If given it should contain the first lines of help text output.
Otherwise 
.B "usage: PROG [options]"
will be printed, where PROG is the string given as the 
.B "prog" 
argument, see above.
.TP 
.B "uogetopt_t *" 
is an array 
and describes the options the program recognizes. It's described in detail below.
.TP
.B "const char *tail"
is optional. If given it should contain the last lines of help text output, to be
printed after the option have been described.

.SH "UOGETOPT_T"
A uogetopt_t describes a single option, and contains
the following members:

.TP 
.B "char shortname"
is the short, one charater, form of the option. 
.I '\\0' 
means it does not have this.
.TP
.B "const char *longname"
is a long option, without leading dashes. 
.I NULL
tells the function that this option does not have a long form.
.TP
.B "int argtype"
describes the type of the variable to hold the option in. See 
.IR "VARIABLE TYPES" ,
below.
.TP
.B "void *var"
points to a place to store the option value into. 
.TP
.B "int val"
will be used as a value for options with flag character.
.TP
.B "const char *help"
is the help text for this option. See 
.IR "HELP TEXTS" ,
below.
.TP
.B "const char *paraname"
is a name for the parameter, to be used in the help text.

.SH "VARIABLE TYPES"
The 
.I argtype 
member of the uogetopt_t structure describes the type of an argument. 
The are 6 possible variable types now:
.TP
.B "UOGO_FLAG"
is a simple flag. The value of the 
.I "val" 
member will be stored into the 
.I "var" 
member, which has to point to an integer type variable (int). Use it like this:
.EX
.nf
int yesno=-1;
[other code]
{'y',"yes", UOGO_FLAG,&yesno,1,"`yes' help text",0},
{'n',"no",  UOGO_FLAG,&yesno,0,"`no' help text",0},
.fi
.EE
.TP
.B "UOGO_FLAGOR"
is like UOGO_FLAG, except that does not overwrite the memory 
.I "var" 
points to, but 
.I ORs 
the
.I "val"
into that.
.TP
.B "UOGO_STRING"
stores the address of the option value string in the memory
.I "var"
points to. 
.I "val"
is unused in the case. Please note that the memory for the string 
is not dynamically allocated. Therefore it's unsafe to change the
.I "char **argv"
array given to the uogetopt function. Use it like this:
.EX
.nf
char *inputfilename;
[other stuff]
{'i',"input", UOGO_STRING,&inputfilename,0,
 "Name of the input file","FILENAME"},
.fi
.EE

.TP
.B "UOGO_ULONG"
converts the option value into an unsigned long and stores it into
the memory 
.I "var" 
points to (which should be of type 
.IR "unsigned long" ).

The options value may be written in hexadecimal (0x...), octal
(0755) or plain decimal. Use it like this:

.EX
.nf
unsigned long ul;
[other stuff]
{'m',"maxsize", UOGO_ULONG,&ul,0,
 "Maximal size of something","BYTES"},
.fi
.EE

The 
.I "val" 
member is ignored.

.TP
.B "UOGO_LONG"
is almost identical to the 
.I "UOGO_ULONG"
type except that it handles signed longs.

.TP 
.B "UOGO_CALLBACK"
is used for other data types or special things. If the 
.I uogetopt 
functions happens to meet such an option it calls the callback function to handle the value.
This is of course meant to be the primary way to extend the uogetopt datatypes.
Use it like this:
.EX
.nf
int from;
int to;

/* error checking omitted */
static int 
handle_range(uogetopt_t *opt, const char *val)
{
    const char *colon=strchr(val,':');
    if (!colon) {
        from=INT_MIN;
        to=(int)strtol(val,NULL,0);
    } else {
        from=(int)strtol(val,NULL,0);
        if (!colon[1])
          to=INT_MAX;
        else
          to=(int)strtol(colon+1,NULL,0);
    }
    return 0; /* ok */
}
[other stuff]
{'l',"loop-boundaries", UOGO_CALLBACK,handle_range,0,
 "Boundaries of the loop in FROM:TO notation","RANGE"},
.fi
.EE
.TP 
.B "UOGO_PRINT"
just print the help text if this option occurs (this
can be used for e.g. copyright options).
.TP 
.B "UOGO_HIDDEN"
This is not really a type, but a value to be ORed with the
type, and make uogetopt not print this option in --help or --longhelp lists
(this may be useful for maintainer mode options).
.EX
.nf
int risky=0;
#define PACKAGE "my package"
[other code]
{ 0 ,"buggy-new-option", UOGO_FLAG|UOGO_HIDDEN,&risky,1,"`yes' help text",0},
{ 0 ,"copyright", UOGO_PRINT|UOGO_HIDDEN,0,0,
  "Copyright (C) 1999 Uwe Ohse\\n"
  PACKAGE " comes with NO WARRANTY,\\n"
  "to the extent permitted by law.\\n"
  "You may redistribute copies of " PACKAGE "\\n"
  "under the terms of the GNU General Public License.\\n"
  "For more information about these matters,\\n"
  "see the files named COPYING.\\n",0},
.fi
.EE
(the --copyright option is used by the --version stuff. It's not needed to 
print it's output in every --longhelp text, so i marked it as hidden.)
.SH "HELP TEXTS"
One of the more important reasons to write uogetopt was to get better
support for help texts. I often forgot to update the help texts when
i changed or added some option (of course even more often i forgot to
update the manual, but i didn't see a good way for a getopt function to
fix that...).

The first thing printed is the

.SS "help text header"
The 
.B head
argument to the 
.B uogetopt 
function will be printed at the top of the help text. If this has
been omitted then
.B "usage: PROG [options]"
will be printed. 

Then follows, after an empty line, the
.SS "help texts for options"
which is taken from the options description given to the
uogetopt function, it's the 
.B help 
member of the 
.I "uogetopt_t"
variable.
In it's basic form is a single string, without a line feed in it. This will
be printed upon request for the long or short help text of an option.
Longer help texts are possible. Multiple lines must be separated by line feeds.
The first line is used for short and long help texts, the following lines are
used for long help only. 

In the long help text is requested the full text will be printed with correct
indentation. An additional line feed is printed after every long help text
(this makes it easy to include empty lines in the output which serve as 
separators).

The 
.B paraname
member of the uogetopt_t is optional, but should contain a name for the
parameter. If it's missing 
.B uogetopt
will just print a name for the variable type.

Examples:
.br
.nf
   {'i',"input", UOGO_STRING,&inputfilename,0,
    "Name of the input file\\n"
    "This file should probably better exist.\\n"
    "If this option is omitted then stdin will be used","FILENAME"},
   {'o',"output", UOGO_STRING,&outputfilename,0,
    "Name of the output file\\n"
    "If this option is omitted then stdout will be used\\n","NULL"},
   {'u',"user", UOGO_STRING,&username,0,
    "Name of the user\\n","USERNAME"},
   {'x',"exit", UOGO_FLAG,&exit_flag,0,
    "Exit without doing anything","FILENAME"},
.fi
.br
results in the following output for
.BR --help :
.br
.nf
     -i, --input=FILENAME       Name of the input file
     -o, --output=STRING        Name of the output file
     -u, --user=USERNAME        Name of the user
     -x, --exit                 Exit without doing anything
.fi
.br
and in the following output for
.BR --longhelp :
.br
.nf
     -i, --input=FILENAME       Name of the input file
                                This file should probably better exist.
                                If this option is omitted then stdin will be used
     -o, --output=STRING        Name of the output file
                                If this option is omitted then stdout will be used

     -u, --user=USERNAME        Name of the user
     -x, --exit                 Exit without doing anything
.fi
.br

The last part of the help text is the
.SS "help text footer"
which is either empty (NULL) or contains some text which will be printed
unindented.

.SS "long help for single options"
If --help or --longhelp options are used to see the help text for a single
option the long help will be printed as described above, but without any
header or trailer.

.SS "version help"
The --version option will print version and, optionally, copyright information, 
which will be taken from the help text of the 
.B --copyright 
option, in case you provided that.

.SH "IMPLICIT OPTIONS"
.I uogetopt
implements some options by itself. These are:
.TP
.B "\-\-uogetopt-paranoid"
Print a help text as if the 
.B "--longhelp" 
options would have been given instead, but prints out a warning if the
help text does not fit on 80 character wide screens. This is not
implemented very sophisticated.
.TP
.B "\-\-help [OPTION]"
Print the first line of the help text of all options, or the
long help for the named option 
.IR OPTION ,
which may be given as a short or long option with or without
leading dashes.
.TP
.B "\-\-longhelp [OPTION]"
Shows longer help text for all or one option. 
.B "\-\-longhelp"
without an argument shows the long help for this
option, without arguments it shows the long description
for all options.
.TP
.B "\-\-version"
prints version information, consisting of the
.IR "prog" ,
.IR "package" and
.I "version" 
information provided to uogetopt function. It also prints
copyright information if available.
.TP
.B "implicit short options"
If 
.I uogetopt
sees a 
.B "\-h" 
or
.B "\-?"
option and the character is not already handled by a short option
then 
.i uogetopt
will treat that option like
.BR "\-\-help" .
Likewise 
.B "\-v" 
or 
.B "\-V" 
will be mapped to 
.BR "\-\-version" . 
These mappings do 
.B not
occur if the short option is already in use.


.SH ERROR HANDLING
uogetopt() will exit upon errors.  Of course this could be changed.

.SH COPYRIGHT
Copyright 1998,1999 Uwe Ohse.
.br
This routines are published under the GNU
General Public License,
see
.IR COPYING 
for a full statement of rights and permissions.
.SH AUTHOR
Uwe Ohse


NEWS
0.2.4 - 1 November 1999

* some cleanup and minor bug fixes
* utftpd left SCCS lock files laying around in case of a failed put
  request.

0.2.3 - 5 September 1999

* the server now ignores garbage bytes at the end of a packet, unless
  there were tftp options.
* "must not write as root" check moved some lines up, to check against
  this before we create files (old version left 0 length files).

0.2.2 - 22 August 1999

* the server now uses adaptive timeout.
* the server now unlinks the received file if it was not received 
  completely.
* the server now includes a workaround for the 
  "Sorcerer's Apprentice Syndrome" (as it's called by RFC 1123)

0.2.1 - 2 May 1999

* rewrote configuration file handling. It's now much more 
  beautiful. 
  See sample.config for more information.
  The old stuff partially continues to work, except that
* the \ at the end of line (to split long lines into smaller pieces)
  will not work anymore. Sorry for that, but it's just too much
  pain to fix this.

0.2.0 - 2 May 1999
* understands a new experimental tftp option (revision).
* major internal interface cleanup.
* now reads version controlled files from a pipe. This means that
  it now can serve read request even when the same file is just uploaded.
* "lying around" copies of version controlled files will now be ignored.
* -l option (GNU inetutils) is now recognized (and ignored,
  utftpd logs by default).
* -n option (GNU inetutils) implemented. It's used to suppress NAK
  answers for "file not found". 
* Can now operate without a configuration file, in a mode largely 
  compatible with the traditional UNIX tftp server: Access control
  based on a list of directories given on the command line and the
  UNIX access bits (read/write for others). Therefore
* it may now be used as a drop-in replacement for GNU inetutils tftpd.

0.1.5 - 10 April 1999
* added RCS support. See the utftpd.conf manual page for a description 
  of the new configuration variables rcs-ci and rcs-co.
* improved SCCS support: unchecked files will not be checked in 
  (that's useful if your router saves his configuration regulary even 
  if it's unchanged). See the utftpd.conf manual page for a 
  description of the new configuration file variables sccs-clean and
  sccs-unget.
* improved checksuite to check SCCS/RCS handling.
* fixed some small bugs.
* improved compatibility (SCO 3.2v4.2, BSDI 3.1, Solaris 2.6 and 2.5.1,
  OpenBSD 2.4/alpha, HPUX 9.00/m68k).

0.1.4 - 5 April 1999
* configuration (cdb) file may now reside outside the chrooted area.
* understands RFC 2348 "blksize" option.
* understands RFC 2349 "timeout" option (but doesn't know about "tsize" 
  now).
* test suite added.
* documentation fixes.
* server now knows about wildcards.
* threw out some bugs and unused library files.
* there's now a client program (utftp).

0.1.3 - 28 March 1999
* uh, 0.1.2 didn't work because of those stupid last-moment-large-rewrite ... 
  Ok, that's why i tested it on my companies tftp server.
* smaller log things.

0.1.2 - 28 March 1999
* manual pages
* rpm fixed

0.1.1 - 21 March 1999
* put on cvs


uogetopt.c
/* uogetopt.c: somewhat GNU compatible reimplementation of getopt(_long) */

/*
 * Copyright (C) 1999 Uwe Ohse
 * 
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 * 
 * As a special exception this source may be used as part of the
 * SRS project by CORE/Computer Service Langenbach
 * regardless of the copyright they choose.
 * 
 * Contact: uwe@ohse.de
 */

/*
 * main differences:
 * + new feature: --version 
 * + new feature: --help  (standard help text, one line per option)
 * + new feature: --help OPTION (maybe somewhat more)
 * + new feature: --longhelp (all --help OPTION)
 * - no "return in order"
 * o really no support for i18n.
 * o small code, about 65% of GNU C library stuff.
 * o it doesn't do black magic, like that GNU stuff.
 *
 * Note that the following statement from the GNU getopt.c was the cause
 * for reinventing the wheel:
 *    Bash 2.0 puts a special variable in the environment for each
 *    command it runs, specifying which ARGV elements are the results of
 *    file name wildcard expansion and therefore should not be
 *    considered as options.
 * i decided that this wheel is to broken to be reused. Think of that
 * "-i" trick. As time passed by i calmed down, but now uogetopt is
 * better than GNU getopt ...
 */

#include "config.h"
#include <string.h>
#include <errno.h>
#include <unistd.h> /* write */
#include <stdlib.h> /* getenv */
#include "uogetopt.h"
#include "str2num.h"

static int paranoid=0;


#define L longname
#define S shortname
#define LLEN 80
#define OFFSET 29 /* keep me < LLEN */

void 
uogetopt_out(int iserr,const char *s)
{
	int fd=1;
	if (iserr)
		fd=2;
	write(fd,s,strlen(s));
}

static void 
uogetopt_outn(void (*out)(int iserr,const char *),
	int iserr,const char *s,const char *t,const char *u,const char *v)
{ out(iserr,s); if (t) out(iserr,t); if (u) out(iserr,u); if (v) out(iserr,v); }
#define O2(E,s,t) do { uogetopt_outn(out,(E),(s),(t),0,0); } while (0)
#define O3(E,s,t,u) do { uogetopt_outn(out,(E),(s),(t),(u),0); } while (0)
#define O4(E,s,t,u,v) do { uogetopt_outn(out,(E),(s),(t),(u),(v)); } while (0)

static void
uogetopt_hlong(void (*out)(int iserr,const char *),const char *argv0, const char *s,uogetopt_t *o)
{
	int len;
	long l;
	len=str2long(s,&l,0);
	if (len<=0) { O4(1,argv0,": ",s," is not an signed long\n");_exit(2);} 
	*(long *)o->var=l;
}

static void
uogetopt_hulong(void (*out)(int iserr,const char *),const char *argv0, const char *s,uogetopt_t *o)
{
	int len;
	unsigned long ul;
	len=str2ulong(s,&ul,0);
	if (len<=0) { O4(1,argv0,": ",s," is not an unsigned long\n");_exit(2);} 
	*(unsigned long *)o->var=ul;
}

static void
uogetopt_describe(void (*out)(int iserr,const char *),uogetopt_t *o, int full)
{
/* 
  -s, --size                 print size of each file, in blocks
  -S                         sort by file size
      --sort=WORD            ctime -c, extension -X, none -U, size -S,
123456789012345678901234567890
I don't really know for what the spaces at the start at the line are
for, but i suppose someone will need them ...
*/
	int l;
	const char *p;
	char *q;
	char buf[LLEN+1];
	char minbuf[2];
	out(0,"  ");
	if (o->S) { buf[0]='-'; buf[1]=o->S; buf[2]=0; out(0,buf); } /* -X */
	else out(0,"  ");

	if (o->S && o->L) out(0,", --");
	else if (o->L)    out(0,"  --");
	else              out(0,"    ");

	l=8;

	if (o->L) {l+=strlen(o->L);out(0,o->L);}

	if (o->argtype!=UOGO_FLAG && o->argtype!=UOGO_FLAGOR) {
		if (o->L) out(0,"=");
		else      out(0," ");
		l++;
		if (o->paraname) {out(0,o->paraname);l+=strlen(o->paraname);}
		else if (o->argtype==UOGO_STRING) {out(0,"STRING");l+=6;}
		else {out(0,"NUMBER");l+=6;}
	}
	/* fill up with spaces (at least one) */
	q=buf; do { *q++=' '; l++; } while (l<OFFSET); *q=0; 
	out(0,buf);
	if (paranoid && l>29) out(0,"\n###time for shorter options?\n");

	/* 1. line of help */
	if (!o->help) {out(0,"???\n"); return; }
	p=strchr(o->help,'\n');
	if (!p || !p[1]) { 
		out(0,o->help); 
		if (!p) out(0,"\n");
		if (paranoid && strlen(o->help)>=51) out(0,"###the last help text is too long for 80 characters screens.\n");
		return;
	}
	minbuf[1]=0;
	p=o->help;
	do {minbuf[0]=*p;out(0,minbuf); } while (*p++!='\n');
	if (paranoid && p-o->help>=51) out(0,"###the last help text is too long for 80 characters screens.\n");

	if (!full) return;
	memset(buf,' ',OFFSET); buf[OFFSET]=0;
	do { 
		const char *r;
		out(0,buf);
		r=p;
		do {minbuf[0]=*p++;out(0,minbuf);} while (*p && p[-1]!='\n');
		if (paranoid && p-r>=51) out(0,"###the last help text is too long for 80 characters screens.\n");
		if (!*p) out(0,"\n");
	} while (*p && p[1]);
}

static void	
uogetopt_printver(void (*out)(int iserr,const char *),
	const char *prog, const char *package, const char *version)
{
		out(0,prog);
		if (package) { O3(0, " (",package,")"); }
		O3(0," ",version,"\n");
}

static int dummy;
static uogetopt_t hack_longhelp=
    {0,"longhelp",   UOGO_FLAG,&dummy,0,
	     "Show longer help texts for all or one variable\n"
       /* 12345678901234567890123456789012345678901234567890" */
		 "--longhelp with an argument shows the long help\n"
		 "for this option, without arguments it shows the\n"
		 "long description of all options.","[OPTION-NAME]"
	};
static uogetopt_t hack_help=
    {0,"help",   UOGO_FLAG,&dummy,0,
       /* 12345678901234567890123456789012345678901234567890" */
		 "Show a list of options or the long help on one.\n"
		 "--help with an argument shows the long helptext\n"
		 "of that option, without an argument it will list\n"
		 "all options.","[OPTION-NAME]"
	};
static uogetopt_t hack_version=
    {0,"version",   UOGO_FLAG,&dummy,0,
	     "Show version information.",0
	};

static char minus_help[]="--help";
static char minus_version[]="--version";
void 
uogetopt(const char *prog, const char *package, const char *version, 
	int *argc, char **argv, void (*out) P__((int iserr,const char *)), 
	const char *head, uogetopt_t *opts, const char *tail)
{
	int i;
	int posix;
	int newargc;
	int ocount;
	int is_longhelp;
	int h_used=0;
	int v_used=0;
	int ques_used=0;
	int V_used=0;
	uogetopt_t *copyright=0;
	if (!argv[1]) return;
	if (0==strcmp(argv[1],"--uogetopt-paranoid")) 
		{ union { char *c; const char *cc; }d; d.cc="--longhelp"; argv[1]=d.c; paranoid=1; }
	for (i=0;opts[i].S || opts[i].L;i++) {
		if (!opts[i].var && !(opts[i].argtype&UOGO_HIDDEN)) {
			/* no use to waste code for detailed error messages, this only
			 * happens during development.
			 */
			out(1,"NULL variable address in uogetopt()\n");
			_exit(1);
		}
		if (opts[i].L && 0==strcasecmp(opts[i].L,"copyright"))
			copyright=&opts[i];
		switch (opts[i].S) {
		case 'v': v_used=1; break;
		case 'h': h_used=1; break;
		case 'V': V_used=1; break;
		case '?': ques_used=1; break;
		}
	}
	/* try to map -?, -h to --help, -V, -v the --version */
	if (!argv[2] && argv[1][0]=='-') {
		if (argv[1][1]=='h' && !h_used) argv[1]=minus_help;
		if (argv[1][1]=='?' && !ques_used) argv[1]=minus_help;
		if (argv[1][1]=='v' && !v_used) argv[1]=minus_version;
		if (argv[1][1]=='V' && !V_used) argv[1]=minus_version;
	}
	ocount=i;
	if (argv[1] && 0==strcmp(argv[1],"--version")) {
		uogetopt_printver(out,prog,package,version);
		if (copyright && copyright->help) {
			out(0,"\n");
			out(0,copyright->help);
		}
		_exit(0);
	}
	is_longhelp=(0==strcmp(argv[1],"--longhelp"));
	if (argv[1] && (is_longhelp || 0==strcmp(argv[1],"--help"))) {
		if (argv[2]) { 
			uogetopt_t *u=0;
			if (argv[2][0]=='-') argv[2]++;
			if (argv[2][0]=='-' && argv[2][1]) argv[2]++;
			for (i=0;i<ocount;i++) {
				if (opts[i].L && 0==strcmp(opts[i].L,argv[2])) break;
				if (opts[i].S && !argv[2][1] && argv[2][0]==opts[i].S) break;
			}
			if (i!=ocount) u=&opts[i]; 
			else if (0==strcmp(argv[2],"--longhelp") || 0==strcmp(argv[2],"longhelp")) u=&hack_longhelp;
			else if (0==strcmp(argv[2],"--help") || 0==strcmp(argv[2],"help")) u=&hack_help;
			else if (0==strcmp(argv[2],"--version") || 0==strcmp(argv[2],"version")) u=&hack_version;
			if (!u) { (out)(0,"no such option\n"); _exit(1); }
			if (!u->help) { (out)(0,"no help available\n"); _exit(1); }
			uogetopt_describe(out,u,1); /* long help every time */
			_exit(0);
		}
		/* uogetopt_printver(out,prog,package,version); it's against the GNU standards */
		if (head) out(0,head); 
		else { O3(0,"usage: ",prog," [options]\n");}
		out(0,"\n");
		for (i=0;i<ocount;i++) {
			if (!(opts[i].argtype & UOGO_HIDDEN))
				uogetopt_describe(out,&opts[i],is_longhelp);
		}
		uogetopt_describe(out,&hack_version,is_longhelp);
		uogetopt_describe(out,&hack_help,is_longhelp);
		uogetopt_describe(out,&hack_longhelp,is_longhelp);
		if (tail) {out(0,tail); out(0,"\n");}
		_exit(0);
	}

	posix=!!getenv("POSIXLY_CORRECT");
	newargc=1;
	for (i=1;i<*argc;i++) {
		if (*argv[i]!='-') {
			if (posix) { 
		  copyrest:
				while (argv[i]) argv[newargc++]=argv[i++];
				argv[newargc]=0;
				*argc=newargc;
				return;
			}
			argv[newargc++]=argv[i];
			continue;
		}
		if (argv[i][1]=='-') { 
			int j;
			int ioff=1;
			char *para;
			uogetopt_t *o;
			int at;
			para=strchr(argv[i],'=');
			if (para) { *para++=0;ioff=0;}
			else {para=argv[i+1];}
			if (!argv[i][2]) { i++; goto copyrest; }
			o=opts;
			for (j=0;j<ocount;o++,j++)
				if (o->L && 0==strcmp(o->L,argv[i]+2)) break;
			at=o->argtype & (~(UOGO_HIDDEN));
			if (j==ocount || at==UOGO_NOOPT) {O4(1,argv[0],": illegal option -- ",argv[i],"\n");_exit(2);}
			if ((at==UOGO_FLAG || at==UOGO_FLAGOR || at==UOGO_PRINT) && ioff==0) /* --x=y */
				{ O4(1,argv[0],": option doesn't allow an argument -- ",argv[i],"\n");_exit(2);}
			if (at==UOGO_FLAG) { *((int*)o->var)=o->val; continue;}
			if (at==UOGO_FLAGOR) { *((int*)o->var)|=o->val; continue;}
			if (at==UOGO_PRINT) { out(0,o->help); if (o->help[strlen(o->help)-1] != '\n') out(0,"\n"); _exit(0); }
			if (!para) {O4(1,argv[0],": option requires an argument -- ",argv[i]+2,"\n");_exit(2);}
			if (at==UOGO_STRING) { *((char **)o->var)=para; i+=ioff; continue;}
			if (at==UOGO_ULONG) { uogetopt_hulong(out,argv[0],para,o); i+=ioff; continue; }
			if (at==UOGO_LONG) { uogetopt_hlong(out,argv[0],para,o); i+=ioff; continue; }
			if (at==UOGO_CALLBACK) { 
				union disqualify {
					void *v;
#if defined(PROTOTYPES) || defined(__cplusplus)
					void (*fn)(uogetopt_t *,const char *);
#else
					void (*fn)();
#endif
				} dq;
				dq.v=o->var;
				(*dq.fn)(o,para);
				i+=ioff; continue;
			}
			O4(1,argv[0],": internal problem, unknown option type for `",argv[i],"'\n");
			_exit(2);
		} else { 
			int j;
			for (j=1;argv[i][j];j++) { /* all chars */
				char c=argv[i][j];
				int k;
				int nexti;
				char *p;
				char optstr[2];
				uogetopt_t *o;
				int at;
				optstr[0]=c;optstr[1]=0;
				o=opts;
				for (k=0;k<ocount;k++,o++) {
					if (o->S && o->S==c) 
						break;
				}
				at=o->argtype & (~(UOGO_HIDDEN));
				if (k==ocount || at==UOGO_NOOPT) {
					O4(1,argv[0],": illegal option -- ",optstr,"\n");
					_exit(2);
				}
				if (at==UOGO_FLAG) { *((int*)o->var)=o->val; continue;}
				if (at==UOGO_FLAGOR) { *((int*)o->var)|=o->val; continue;}
				if (at==UOGO_PRINT) { out(0,o->help); if (o->help[strlen(o->help)-1] != '\n') out(0,"\n"); _exit(0); }
				/* options with arguments, first get arg */
				nexti=i; p=argv[i]+j+1; 
				if (!*p) { p=argv[i+1];
					if (!p) {O4(1,argv[0],": option requires an argument -- ",optstr,"\n");_exit(2);}
					nexti=i+1; }
				if (at==UOGO_STRING) { *((char **)o->var)=p; i=nexti; break; }
				if (at==UOGO_ULONG) { uogetopt_hulong(out,argv[0],p,o); i=nexti; break; }
				if (at==UOGO_LONG) { uogetopt_hlong(out,argv[0],p,o); i=nexti; break; }
				if (at==UOGO_CALLBACK) { 
					union disqualify {
						void *v;
#if defined(PROTOTYPES) || defined(__cplusplus)
						void (*fn)(uogetopt_t *,const char *);
#else
						void (*fn)();
#endif
					} dq;
					dq.v=o->var;
					(*dq.fn)(o,p);
					i=nexti; break;
				}
				O4(1,argv[0],": internal problem, unknown option type for `",optstr,"'\n");
				_exit(2);
			}
		}
	}
	*argc=newargc;
	argv[newargc]=0;
}

#ifdef TEST

int flag=0;
char *string;
unsigned long ul=17;
signed long lo=18;
uogetopt_t myopts[]={
	{'t',"test",UOGO_FLAG,&flag,1,"test option\n"},
	{'l',"longtest",UOGO_FLAGOR,&flag,2,"test option with some characters\nand more than one line, of which one is really long, of course, just to test some things, like crashs or not crash, life or 42\n"},
	{'L',0,UOGO_FLAG,&flag,2,"test option with some characters\nand more than two lines\nwith not much information\nbut a missing newline at the end: 43"},
	{'M',0,UOGO_FLAG,&flag,2,"test option with some characters\nand two lines: 44\n"},
	{'N',0,UOGO_FLAG,&flag,2,"test option with some characters\nand two lines, no nl at end: 45"},
	{'u',"ulong",UOGO_ULONG,&ul,2,"set a ulong variable\n","Ulong"},
	{'n',"nulong",UOGO_LONG,&lo,2,"set a long variable\n","long"},
	{'s',"string",UOGO_STRING,&string,2,"set a string variable\n","String"},
	{0,0}
};
int 
main(int argc, char **argv)
{

	printf("old argc=%d\n",argc);
	uogetopt("uogetopt","misc","0.1",
		&argc, argv,
		writeerr,
		"head\n",myopts,"tail\n");
	printf("flag=%d\n",flag);
	printf("string=%s\n",string ? string : "NULL");
	printf("ulong=%lu\n",ul);
	printf("long=%ld\n",lo);
	printf("new argc=%d\n",argc);
	{ int i; for (i=0;argv[i];i++) printf("argv%d: %s\n",i,argv[i]);}
	exit(0);
}
#endif
